### ISSUE TITLE
Build/release Python 3.10 tflite-runtime wheels to PyPI

### ISSUE THREAD

**Issue by johnthagen** [tensorflow/tensorflow#56137]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Build/Install

### Source

binary

### Tensorflow Version

2.9.0

### Custom Code

No

### OS Platform and Distribution

Linux & macOS

### Python version

3.10

### Current Behaviour?

[Tensorflow 2.9.0 was released with Python 3.10 support](https://pypi.org/project/tensorflow/2.9.0/#files).

We would like to support Python 3.10 in production, but in order to do this we need all of the following packages to support Python 3.10:

- `tensorflow`
- `tensorflow-macos` (for ARM macOS)
- `tflite-runtime` (for stripped down production on Linux)

Neither [`tensorflow-macos`](https://pypi.org/project/tensorflow-macos/#history) or [`tflite-runtime`](https://pypi.org/project/tflite-runtime/) have 2.9.0 releases on PyPI yet.

### Standalone code to reproduce the issue

Create a Python 3.10 virtual environment on Linux:

```bash
(venv) $ python -m pip install tflite-runtime
```

Create a Python 3.10 virtual environment on macOS ARM:

```bash
(venv) $ python -m pip install tensorflow-macos
```

### Relevant log output

_No response_</details>

**Comment by SoroushMB** [comment]
Hey How u doing
I think you should use mini conda environment for installing Tensorflow
Also,update your GCC and G++ compiler and learn more about your hardware.
It was my idea and I think it will help you out.

**Comment by johnthagen** [comment]
@SoroushMB We are interested in installing pre-compiled wheels from PyPI rather than trying to build Tensorflow/Tensorflow Lite on each end system.

**Comment by snowuyl** [comment]
It seems TensorFlow macOS 2.9.0 is not release yet. When will it release?
pip3 install tensorflow-macos==2.9.0
DEPRECATION: Configuring installation scheme with distutils config files is deprecated and will no longer work in the near future. If you are using a Homebrew or Linuxbrew Python, please see discussion at https://github.com/Homebrew/homebrew-core/issues/76621
ERROR: Could not find a version that satisfies the requirement tensorflow-macos==2.9.0 (from versions: 2.5.0, 2.6.0, 2.7.0, 2.8.0)
ERROR: No matching distribution found for tensorflow-macos==2.9.0

**Comment by johnthagen** [comment]
@sachinprasadhs I see that `tensorflow-macos` 2.9.0 was released, but there is not a 3.10 wheel:

- https://pypi.org/project/tensorflow-macos/2.9.0/#files

**Comment by sachinprasadhs** [comment]
Hi, The release of `tensorflow-macos` is done by apple, you can raise issue in https://developer.apple.com/forums/. Thanks!

**Comment by johnthagen** [comment]
@sachinprasadhs What about `tflite-runtime`? Thanks.

**Comment by snowuyl** [comment]
% pip3 install tensorflow-macos==2.9.0
Successfully installed tensorflow-macos-2.9.0
% python3
Python 3.9.12 (main, May  8 2022, 17:57:49) 
[Clang 13.1.6 (clang-1316.0.21.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
>>> print(tf.__version__)
2.9.0
>>> from tensorflow.keras import layers

**Comment by kulinseth** [comment]
> @sachinprasadhs / @kulinseth I see that `tensorflow-macos` 2.9.0 was released, but there is not a 3.10 wheel:
> 
> * https://pypi.org/project/tensorflow-macos/2.9.0/#files

We have planned for py38/py39 releases so far. We can look into adding 3.10 support.

**Comment by sachinprasadhs** [comment]
> @sachinprasadhs What about `tflite-runtime`? Thanks.

It will be released in a week or so.

**Comment by terryheo** [comment]
`tflite-runtime 2.9.1` also has been released.
https://pypi.org/project/tflite-runtime/2.9.1

**Comment by johnthagen** [comment]
@terryheo Thanks!

We were really hoping for a Python 3.10 wheel (like is available for `tensorflow` and `tensorflow-macos` packages). 

Would it be possible to build and upload Python 3.10 for `tflite-runtime 2.9.1`?

**Comment by terryheo** [comment]
Oh, I see. I'll handle it.

**Comment by johnthagen** [comment]
@terryheo Just checking back in to see if a Python 3.10 wheel could be published. Thanks.

**Comment by johnthagen** [comment]
@terryheo Checking in again. Is there anything the community can help with to get Python 3.10 wheels for `tflite-runtime`? We'd really like to deploy against Python 3.10 in production.

I see the latest [`tflite-runtime-nightly` on PyPI](https://pypi.org/project/tflite-runtime-nightly/2.11.0.dev20220819/#files) still only targets Python 3.7-3.9.

Let us know how we can help. ❤️

**Comment by jakew009** [comment]
Any updates?

**Comment by terryheo** [comment]
Hello, @johnthagen

TFLite is using [this Makefile](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/pip_package/Makefile) to build PyPI package as the following.

```
make -C tensorflow/lite/tools/pip_package docker-build PYTHON_VERSION="3.10" "x86_64" "manylinux2014_x86_64" "tflite-runtime-nightly"
```

The script doesn't work well with `PYTHON_VERSION="3.10"`. It fails on finding necessary Python packages.

```
Reading state information...
E: Unable to locate package libpython3.10-dev:armhf
E: Couldn't find any package by glob 'libpython3.10-dev'
E: Couldn't find any package by regex 'libpython3.10-dev'
```

We need to update the [Makefile](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/pip_package/Makefile)  and the [Dockerfile.py3](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/pip_package/Dockerfile.py3) to release Python 3.10 wheels.

If you could make a PR to fix the issue, I'll be happy to review it.

**Comment by samypr100** [comment]
@terryheo 

It seems the Python 3.10 build for `armhf` is failing upstream in deadsnakes ppa based on the status in the [packages details](https://launchpad.net/~deadsnakes/+archive/ubuntu/ppa/+packages). Here's the [last armhf build](https://launchpad.net/~deadsnakes/+archive/ubuntu/ppa/+build/24544153) attempt failing on `bionic` (ubuntu 18.04).

Focal seems to be working, so might be worth a shot changing `BASE_IMAGE ?= ubuntu:18.04` to `BASE_IMAGE ?= ubuntu:20.04` on the [Makefile](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/pip_package/Makefile) or wait until deadsnakes has a working build.

**Comment by johnthagen** [comment]
> or wait until deadsnakes has a working build

It looks like DeadSnakes PPA won't be able to fix this on `bionic` / 18.04, so moving up to 20.04 is likely the best option to try first:

- https://github.com/deadsnakes/issues/issues/232#issuecomment-1289398337

**Comment by terryheo** [comment]
Using 20.04 can resolve the package installation error. But it fails on numpy build.

```
      x86_64-linux-gnu-gcc: build/src.linux-x86_64-3.10/numpy/core/src/multiarray/scalartypes.c
      numpy/core/src/multiarray/scalartypes.c.src: In function ‘float_arrtype_hash’:
      numpy/core/src/multiarray/scalartypes.c.src:2968:27: error: incompatible type for argument 1 of ‘_Py_HashDouble’
       2968 |     return _Py_HashDouble((double) PyArrayScalar_VAL(obj, @name@));
      In file included from /usr/include/python3.10/Python.h:77,
                       from numpy/core/src/multiarray/scalartypes.c.src:3:
      /usr/include/python3.10/pyhash.h:10:38: note: expected ‘PyObject *’ {aka ‘struct _object *’} but argument is of type ‘double’
         10 | PyAPI_FUNC(Py_hash_t) _Py_HashDouble(PyObject *, double);
            |                                      ^~~~~~~~~~
      numpy/core/src/multiarray/scalartypes.c.src:2968:12: error: too few arguments to function ‘_Py_HashDouble’
       2968 |     return _Py_HashDouble((double) PyArrayScalar_VAL(obj, @name@));
            |            ^~~~~~~~~~~~~~
      In file included from /usr/include/python3.10/Python.h:77,
                       from numpy/core/src/multiarray/scalartypes.c.src:3:
      /usr/include/python3.10/pyhash.h:10:23: note: declared here
         10 | PyAPI_FUNC(Py_hash_t) _Py_HashDouble(PyObject *, double);
            |                       ^~~~~~~~~~~~~~
      numpy/core/src/multiarray/scalartypes.c.src: In function ‘cfloat_arrtype_hash’:
      numpy/core/src/multiarray/scalartypes.c.src:2976:31: error: incompatible type for argument 1 of ‘_Py_HashDouble’
       2976 |     hashreal = _Py_HashDouble((double)
            |                               ^~~~~~~~
            |                               |
            |                               double
       2977 |             PyArrayScalar_VAL(obj, C@name@).real);
            |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

**Comment by johnthagen** [comment]
@terryheo We are attempting to get 3.10 `armhf` builds fixed upstream, which could make the current 18.04 `armhf` builds work for DeadSnakes, and then Tensorflow could use them without moving up to 20.04:

- https://github.com/python/cpython/issues/98671

**Comment by johnthagen** [comment]
The sha3 ARM fix backport was merged into 3.10:

- https://github.com/python/cpython/pull/98677

[This should be part of 3.10.9, scheduled to be released on 2022-12-05](https://peps.python.org/pep-0619/#bugfix-releases).

**Comment by bperseghetti** [comment]
@johnthagen any downside to just bumping it to 22.04 and give extra headroom on LTS EOL?

**Comment by johnthagen** [comment]
@bperseghetti The downside is that the wheels will be built against a certain GLIBC version. This version will be the minimum GLIBC version required for the wheels to work.

For example,

If the wheels are built within a 22.04 container image, then they will not run properly on 20.04, which this project will want to support since [20.04 is still in active support for another 2 years](https://endoflife.date/ubuntu).

The `manylinux` standards have discussion about this:

- https://github.com/pypa/manylinux

**Comment by bperseghetti** [comment]
@johnthagen are you on the TF Discord? I just joined it and will post in bugs-and-issues. I did change the make file to build 3.10 through docker on 22.04 for arm64. I did it by changing the numpy dependency. How tightly coupled is TF to numpy version?

![image](https://user-images.githubusercontent.com/10233412/209980599-c0f71598-5260-4e1e-b0d0-ef32ce942de2.png)

**Comment by johnthagen** [comment]
> are you on the TF Discord?

@bperseghetti I am not.

Someone from the Tensorflow team would be best at answering this question such as @terryheo who maintains `tflite-runtime`.

**Comment by bperseghetti** [comment]
So, building it from 20.04 with deadsnakes for 3.10 fails, even when trying anything greater than the minimum matching python3.10 capable numpy:

![image](https://user-images.githubusercontent.com/10233412/209984641-be23e35e-5e20-4c74-a76e-21785b3fb2db.png)

**Comment by bperseghetti** [comment]
Honestly something is better than nothing, so I think I might submit a PR for the minor modifications needed to run a wheel for 3.10 on 22.04. This at least unblocks those on the latest LTS. I'm curious how many people use python3.10 on anything running less than glibc2.35.

**Comment by johnthagen** [comment]
> I'm curious how many people use python3.10 on anything running less than glibc2.35.

The current [`3.10` Docker images](https://hub.docker.com/_/python) on DockerHub are based off of either Debian bullseye or buster. I haven't checked the exact glibc versions, but:

[Bullseye ~= Ubuntu 20.04
Buster ~= Ubuntu 18.04](https://askubuntu.com/questions/445487/what-debian-version-are-the-different-ubuntu-versions-based-on)

So wheels built for Ubuntu 22.04 wouldn't be able to run in either of the default Debian-based `python` containers, which is probably a large amount of users.

**Comment by bperseghetti** [comment]
Looks like it won't build on those as they use 3.10.9

python:3.10-buster fails:
![image](https://user-images.githubusercontent.com/10233412/209990270-28f1b2fd-96e2-440c-a94a-513e4ea3c83e.png)


python:3.10-bullseye fails:
![image](https://user-images.githubusercontent.com/10233412/209990199-78b3e93f-e8ab-478d-8c36-762371977b6e.png)

**Comment by bperseghetti** [comment]
My 22.04 built one works on my imx8mplus running 22.04. So, I'm happy.

**Comment by bperseghetti** [comment]
If anyone is in my boat you can use: https://github.com/rudislabs/tensorflow/tree/pr-python310-glibc2.35

To build:

1.  `cd <path>/tensorflow/tensorflow/lite/tools/pip_package` 
2.  `make BASE_IMAGE=ubuntu:22.04 PYTHON_VERSION=3.10 TENSORFLOW_TARGET=aarch64 docker-build`. 

Results in the attached wheel and tested with [NavQPlus](https://nxp.gitbook.io/8mpnavq/). 

Note the wheel needs unzipped as I cannot add a .whl file.

[tflite_runtime-2.12.0-cp310-cp310-linux_aarch64.whl.zip](https://github.com/tensorflow/tensorflow/files/10321885/tflite_runtime-2.12.0-cp310-cp310-linux_aarch64.whl.zip)

**Comment by samypr100** [comment]
Opened a PR #59423 with Option 2 from above just in case

**Comment by samypr100** [comment]
@terryheo Upstream deadsnakes/issues#237 has been fixed, so you shouldn't encounter those `Python.h` errors on your own builds anymore in 20.04. I left my PR open just in case we want to go that route.
